From 0c7fd63349788743332d7b72e8ec962b7e3b9bf4 Mon Sep 17 00:00:00 2001
From: Tim Brown <tim@siftscience.com>
Date: Thu, 19 Jan 2017 14:21:24 -0800
Subject: [PATCH 409/410] HBASE-17658 Fix bookkeeping error with max regions
 for a table

Signed-off-by: tedyu <yuzhihong@gmail.com>
(cherry picked from commit 938aef772d82f7a52228436b141e62af66e11e76)

Change-Id: Ia89c481239346e757515a848787324bf4e1ea86b
Author: Tim Brown <tim@siftscience.com>
Reason: Bug
Ref: CDH-58567
---
 .../hbase/master/balancer/BaseLoadBalancer.java    |    2 +-
 .../master/balancer/TestBaseLoadBalancer.java      |    2 ++
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java
index 3fa14da..48ec1ac 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java
@@ -646,7 +646,7 @@ public abstract class BaseLoadBalancer implements LoadBalancer {
 
       //check whether this caused maxRegionsPerTable in the new Server to be updated
       if (numRegionsPerServerPerTable[newServer][tableIndex] > numMaxRegionsPerTable[tableIndex]) {
-        numRegionsPerServerPerTable[newServer][tableIndex] = numMaxRegionsPerTable[tableIndex];
+        numMaxRegionsPerTable[tableIndex] = numRegionsPerServerPerTable[newServer][tableIndex];
       } else if (oldServer >= 0 && (numRegionsPerServerPerTable[oldServer][tableIndex] + 1)
           == numMaxRegionsPerTable[tableIndex]) {
         //recompute maxRegionsPerTable since the previous value was coming from the old server
diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestBaseLoadBalancer.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestBaseLoadBalancer.java
index 3bdae33..d11b353 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestBaseLoadBalancer.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/master/balancer/TestBaseLoadBalancer.java
@@ -319,6 +319,8 @@ public class TestBaseLoadBalancer extends BalancerTestBase {
 
     // now move region1 from servers[0] to servers[2]
     cluster.doAction(new MoveRegionAction(0, 0, 2));
+    // check that the numMaxRegionsPerTable for "table" has increased to 2
+    assertEquals(2, cluster.numMaxRegionsPerTable[0]);
     // now repeat check whether moving region1 from servers[1] to servers[2]
     // would lower availability
     assertTrue(cluster.wouldLowerAvailability(hri1, servers[2]));
-- 
1.7.9.5

